packages/runtime/src/widgets/__tests__/Conditional.test.tsx

```diff
@@ -0,0 +1,236 @@
+        },
+      ],
+    };
+
+    const [isValid] = hasProperStructure(conditionalProps.conditions);
+    expect(isValid).toBe(false);
+  });
+
+  test("returns false when a condition does not have exactly two properties", () => {
+    const conditionalProps: ConditionalProps = {
+      conditions: [
+        {
+          if: "true",
+          Text: { name: "Text", properties: { text: "Widget A" } },
+          Text2: { name: "Text", properties: { text: "Widget B" } },
+        },
+        {
+          else: null,
+        },
+      ],
+    };
+
+    const [isValid] = hasProperStructure(conditionalProps.conditions);
+    expect(isValid).toBe(false);
+  });
+
+  it('returns false when "if" condition(s) is(are) not the initial condition(s) following "elseif" or "else"', () => {
+    const conditions = [
+      {
+        elseif: "false",
+        Text: { name: "Text", properties: { text: "Widget A" } },
+      },
+      {
+        if: "true",
+        Text: { name: "Text", properties: { text: "Widget B" } },
+      },
+    ];
+
+    const [isValid] = hasProperStructure(conditions);
+    expect(isValid).toBe(false);
+  });
+
+  test("returns false for improper structure", () => {
+    const conditionalProps: ConditionalProps = {
+      conditions: [
+        {
+          if: "false",
+          Text: { name: "Text", properties: { text: "Widget A" } },
+        },
+        {
+          elseif: "true",
+          Text: { name: "Text", properties: { text: "Widget B" } },
+        },
+        {
+          else: null,
+          Text: { name: "Text", properties: { text: "Widget C" } },
+        },
+        {
+          else: null,
+          Text: { name: "Text", properties: { text: "Widget D" } },
+        },
+      ],
+    };
+
+    const [isValid] = hasProperStructure(conditionalProps.conditions);
+    expect(isValid).toBe(false);
+  });
+
+  test("returns true for proper structure", () => {
+    const conditionalProps: ConditionalProps = {
+      conditions: [
+        {
+          if: "false",
+          Text: { name: "Text", properties: { text: "Widget A" } },
+        },
+        {
+          elseif: "true",
+          Text: { name: "Text", properties: { text: "Widget B" } },
+        },
+        {
+          else: null,
+          Text: { name: "Text", properties: { text: "Widget C" } },
+        },
+      ],
+    };
+
+    const [isValid] = hasProperStructure(conditionalProps.conditions);
+    expect(isValid).toBe(true);
+  });
+});
+
+describe("extractWidget Function", () => {
+  it("throws an error for improper structure", () => {
+    const condition = {
+      if: "true",
+    };
+
+    expect(() => extractWidget(condition)).toThrow();
+  });
+
+  it("extracts the correct widget", () => {
+    const condition = {
+      if: "true",
+      Text: { name: "Text", properties: { text: "Widget A" } },
+    };
+
+    const widget = extractWidget(condition);
+    expect(widget.name).toBe("Text");
+  });
```

Comment:
@anserwaseem you have a bug here, the returned widget is malformed (you're unwrapping the widget twice):

```
     {
        name: 'Text',
        properties: { name: 'Text', properties: { text: 'Widget C' } }
      }
```